lxd/apparmor/rsync.go

```diff
@@ -0,0 +1,192 @@
+  capability fowner,
+
+  unix (connect, send, receive) type=stream,
+
+  @{PROC}/@{pid}/cmdline r,
+  @{PROC}/@{pid}/cpuset r,
+  {{ .rootPath }}/{etc,lib,usr/lib}/os-release r,
+
+  {{ .logPath }}/*/netcat.log rw,
+
+  {{ .rootPath }}/run/{resolvconf,NetworkManager,systemd/resolve,connman,netconfig}/resolv.conf r,
+  {{ .rootPath }}/run/systemd/resolve/stub-resolv.conf r,
+
+{{- if .sourcePath }}
+  {{ .sourcePath }}/** r,
+  {{ .sourcePath }}/ r,
+{{- end }}
+
+{{- if .dstPath }}
+  {{ .dstPath }}/** rw,
+  {{ .dstPath }}/ rw,
+{{- end }}
+
+{{- if .snap }}
+  # Snap-specific libraries
+  /snap/lxd/*/lib/**.so* mr,
+{{- end }}
+
+  {{ .execPath }} mixr,
+
+{{if .libraryPath -}}
+  # Entries from LD_LIBRARY_PATH
+{{range $index, $element := .libraryPath}}
+  {{$element}}/** mr,
+{{- end }}
+{{- end }}
+
+  deny {{ .rootPath }}/etc/ssl/openssl.cnf r,
+  deny /sys/devices/virtual/dmi/id/product_uuid r,
+  deny /sys/kernel/mm/transparent_hugepage/hpage_pmd_size r,
+}
+`))
+
+// RsyncWrapper is used as a RunWrapper in the rsync package.
+func RsyncWrapper(sysOS *sys.OS, cmd *exec.Cmd, sourcePath string, dstPath string) (func(), error) {
+	if !sysOS.AppArmorAvailable {
+		return func() {}, nil
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Attempt to deref all paths.
+	if sourcePath != "" {
+		fullPath, err := filepath.EvalSymlinks(sourcePath)
+		if err == nil {
+			sourcePath = fullPath
+		}
+	}
+
+	if dstPath != "" {
+		fullPath, err := filepath.EvalSymlinks(dstPath)
+		if err == nil {
+			dstPath = fullPath
+		}
+	}
+
+	// Load the profile.
+	profileName, err := rsyncProfileLoad(sysOS, sourcePath, dstPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load rsync profile: %w", err)
+	}
+
+	revert.Add(func() { _ = unloadProfile(sysOS, profileName, profileName) })
+	revert.Add(func() { _ = deleteProfile(sysOS, profileName, profileName) })
+
+	// Resolve aa-exec.
+	execPath, err := exec.LookPath("aa-exec")
+	if err != nil {
+		return nil, err
+	}
+
+	// Override the command.
+	newArgs := []string{"aa-exec", "-p", profileName}
+	newArgs = append(newArgs, cmd.Args...)
+	cmd.Args = newArgs
+	cmd.Path = execPath
+
+	// All done, setup a cleanup function and disarm reverter.
+	cleanup := func() {
+		_ = unloadProfile(sysOS, profileName, profileName)
+		_ = deleteProfile(sysOS, profileName, profileName)
+	}
+
+	revert.Success()
+
+	return cleanup, nil
+}
+
+func rsyncProfileLoad(sysOS *sys.OS, sourcePath string, dstPath string) (string, error) {
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Generate a temporary profile name.
+	name := profileName("rsync", uuid.New())
+	profilePath := filepath.Join(aaPath, "profiles", name)
+
+	// Generate the profile
+	content, err := rsyncProfile(sysOS, name, sourcePath, dstPath)
+	if err != nil {
+		return "", err
+	}
+
+	// Write it to disk.
+	err = ioutil.WriteFile(profilePath, []byte(content), 0600)
+	if err != nil {
+		return "", err
+	}
+
+	revert.Add(func() { os.Remove(profilePath) })
```

Comment:
Out of interest, can we just always remove this file via a defer once its been loaded successfully or does it need to linger on disk for the duration the profile is used?

Reply:
There used to be a requirement for the profile to exist on-disk for it to be unloaded, so we've usually kept them around until we're done using them.